### PR TITLE
Binned continuous overlay

### LINKED ISSUE #206: Visualizations: more options for continuous var overlays as binned discrete color maps
should ideally support equal intervals, quantiles and standard deviations. think all vizs w a gradient should gain this feature.

### PR DESCRIPTION
**Pull request by d-callan**
resolves #206

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -1,17 +1,18 @@
-recodeValues <- function(values, desiredValues, dataType) {
-  if (is.null(desiredValues)) return(values)
+recodeValues <- function(values, desiredValues) {
+  if (!length(desiredValues)) return(values)
+  desiredLabels <- unlist(lapply(desiredValues, function(x) {x@binLabel}))
```
Comment (asizemore) [inline R/utils.R:3]: do desired values have to have a binLabel? I might be thinking of a different class, but i thought binLabels were optional? What happens if there is no bin label?

**Review by asizemore — COMMENTED** [review]
had to sign off but will finish first thing tomorrow morning!

File: R/class-plotdata.R [hunk 2]
```diff
@@ -211,6 +203,14 @@ newPlotdata <- function(.dt = data.table(),
   if (!is.null(panel)) { .dt[[panel]] <- updateType(.dt[[panel]], 'STRING') }
   veupathUtils::logWithTime('Base data types updated for all columns as necessary.', verbose)
 
+  if (!is.null(group) && !isOverlayCollection) {
```
[the remark this answers is not in the document]
Comment (d-callan) [reply R/class-plotdata.R:206]: i moved it primarily bc i wanted to update types if necessary before converting the continuous variable to categorical

File: R/utils.R (hunk 1 above)
Comment (d-callan) [reply R/utils.R:3]: all Bins have a binLabel, everything else is optional technically. but there is also validation that requires if a single binStart or binEnd is provided in a BinList, then they all must be.

File: R/utils.R [hunk 3]
```diff
@@ -1,17 +1,18 @@
...
+  # were assuming if we only have binLabels that we should treat the overlay as categorical, regardless of data type
+  breaks <- c(unlist(lapply(desiredValues, function(x) {x@binStart})), desiredValues[[length(desiredValues)]]@binEnd)
+  if (!any(is.na(breaks))) {
+    values <- as.character(cut(values, breaks, labels = desiredLabels, include.lowest = TRUE))
   } else {
-    if (all(unique(values) %in% desiredValues)) return(values)
-    values[!values %in% desiredValues] <- '__UNSELECTED__'
-  }
+    if (all(unique(values) %in% desiredLabels)) return(values)
+    values[!values %in% desiredLabels] <- '__UNSELECTED__'
```
Comment (asizemore) [inline R/utils.R:11]: does this need parentheses? `!(values %in% desiredLabels)`

File: tests/testthat/test-bar.R [hunk 4]
```diff
@@ -764,6 +765,49 @@ test_that("bar() returns appropriately formatted json", {
...
   expect_equal(jsonList$completeCasesTable$variableDetails$variableId, c('int6', 'int11', 'cat6'))
+
+  #binned cont overlay
+  variables <- new("VariableMetadataList", SimpleList(
+    new("VariableMetadata",
+      variableClass = new("VariableClass", value = 'native'),
+      variableSpec = new("VariableSpec", variableId = 'cat6', entityId = 'entity'),
+      plotReference = new("PlotReference", value = 'facet1'),
+      dataType = new("DataType", value = 'NUMBER'),
+      dataShape = new("DataShape", value = 'CONTINUOUS')),
```
Comment (asizemore) [inline tests/testthat/test-bar.R:776]: dataType and dataShape do not match the variable. Same in the test above

File: tests/testthat/test-map-markers.R [hunk 5]
```diff
@@ -245,14 +245,14 @@ test_that("mapMarkers() returns appropriately formatted json", {
...
-      variableSpec = new("VariableSpec", variableId = 'int11', entityId = 'entity'),
+      variableSpec = new("VariableSpec", variableId = 'int7', entityId = 'entity'),
       plotReference = new("PlotReference", value = 'xAxis'),
       dataType = new("DataType", value = 'STRING'),
       dataShape = new("DataShape", value = 'CATEGORICAL'))
   ))
 
   df <- as.data.frame(testDF)
-  df$entity.int11 <- as.character(df$entity.int11)
+  df$entity.int11 <- as.character(df$entity.int7)
```
Comment (asizemore) [inline tests/testthat/test-map-markers.R:255]: should be `df$entity.int7`?

**Review by asizemore — APPROVED** [review]

File: R/utils.R (hunk 3 above)
Comment (d-callan) [reply R/utils.R:11]: i dont think, but for clarity ill add it

**Pull request merged**